Air QRF (combat landing, fastrope) bug fixes

jwoodruff40 opened the pull request:
## What type of PR is this?
- [x] Bug
- [x] Change
- [ ] Feature
- [ ] Miscellaneous

<details><summary><i><b>
Sub-categories:
</b></i></summary>

- [ ] Template
- [ ] Map
- [ ] Config
- [x] Function
- [ ] Localization

</details>

## What have you changed, and why?

**Information:**

> - Minor cleanup and refactor of QRFBehavior
> - Don't fastrope dead units
> - Refactor and reorder of combatLanding. *Should* (hopefully) prevent the most egregious bugs, e.g. failing to unload units altogether, taking off before finishing unload, and / or getting SAD waypoint before unloading units.
> - More robust safe position finding and landing position blacklisting
> - All heli QRFs will attempt combat landing; if landing unsafe, fallback to fastrope

> This *does not* prevent all heli crashes in heavily built-up or vegetated / forested areas, but it's generally a lot better imo.

> I don't seem to get the issues as much in the first place as y'all though this does seem to help, so need others to help test.

**How can your changes be tested, or reproduced?**

> Zeus + `["QRFAIR", Invaders, "attack", 1000, vehicle player, getPosATL player, 1, 0] spawn A3A_fnc_createSupport;`

**Does this PR include changes not authored by you?**

- [x] No
- [ ] Yes (See below)

<details><summary><i><b>
Further info:
</b></i></summary><br>

- [ ] I confirm that I, and by extension this repository, can legally use these third-party changes. (Provide links or author attribution.)

> ...

</details>

## Please verify the following (If possible).

`*` - Mandatory

- [x] These changes are my own, or I have written permission to use them. `*`
- [x] These changes are ready for review, or will be marked as a draft. `*`
- [x] I have loaded the mission in LAN host.
- [ ] I have loaded the mission on a dedicated server.

Is further work needed?

- [x] Needs further testing.
- [ ] Needs further changes.
- [ ] Needs to be converted to a draft.

## Please specify which Issue this PR Resolves (If Applicable).
This PR closes #XYZ.

<hr><details><summary><i><b>
Notes:
</b></i></summary><br>

> ...

</details>

jwoodruff40 committed:
Minor cleanup

jwoodruff40 committed:
Minor refactor to combatLanding

Should prevent qrf vics from starting SAD before unloading troops, failing to unload, and other misc edge cases

jwoodruff40 committed:
Don't fastrope dead units

wersal454 commented on A3A/addons/core/functions/AI/fn_combatLanding.sqf:
```diff
@@ -136,26 +136,32 @@ while {_interval < 0.9999} do
     if ((getPos _helicopter select 2) < 0.25 ) exitwith{_helicopter setdamage 0; sleep 1; _helicopter setdamage _dam;};
 };
 sleep 0.1;
+[_helicopter] call A3A_fnc_smokeCoverAuto;          // Already done by GetOut handler in AIVehInit?
```
yes

jwoodruff40 replied:
Yeah that's not my comment, I just moved that line in the code. Seems to be quite unreliable when it does and doesn't run though.

wersal454 commented on A3A/addons/core/functions/AI/fn_fastrope.sqf:
at least my advanced rappelling extender seems to work)

jwoodruff40 replied:
One thing I was thinking about for 11.9, I wonder if it would be possible to fake a fastrope with sling loading functions.

Like attach an invisible box to units, have the heli sling load attach to that box and lower it to the ground. No idea if that would actually work or not though.

wersal454 replied:
well... I think I know how AR works, so I could try to adopt how it does rope stuff into antistasi. Without porting the mod itself (though it's possible due to MIT licence (though reportedly it has issues on Linux (though in my extender I apparently fixed them)))

wersal454 commented on A3A/addons/core/functions/AI/fn_combatLanding.sqf:
```diff
@@ -165,46 +171,26 @@ _cargoWP2 setWaypointType "SAD";
 _cargoGroup spawn A3A_fnc_attackDrillAI;
 
-
-if(!canMove _helicopter || !alive _driver) exitWith { deleteVehicle _landPad };
```
so you removed "dirty hack" altogether? not sure if it's a good idea. Honestly I would rather propose these changes to the community or at least directly to John or Tiny and ask their opinion.

jwoodruff40 replied:
"Dirty hack" was a Socrates + wersal modification of community's code.

Removing it didn't seem to cause any ill effects for me, but we can re-add in a simplified (more like community) form if helicopter bobbing is an issue.

wersal454 replied:
I'll try looking into changes more indepth tomorrow

wersal454 commented:
More of a general question. Maybe merge this PR and #549 ?

jwoodruff40 commented on A3A/addons/core/functions/AI/fn_combatLanding.sqf:
```diff
@@ -136,26 +136,32 @@ while {_interval < 0.9999} do
+
+deleteVehicle _landPad;
+
+if (count units _cargoGroup isEqualTo 0) exitWith {};
```
This will cause the helicopter + crew to just sit and do nothing if all the cargo group dies but the heli + crew is still alive somehow.

Will fix tomorrow.

jwoodruff40 committed:
QRFBehavior minor refactor;

More robust safe position and blacklist position handling for air vehicles

Attempt combat landing for all helis; fallback to fastrope if terrain objects prevent safe landing

jwoodruff40 commented:
> More of a general question. Maybe merge this PR and #549 ?

I'll have to review 549, but this has made so many changes to fn_fastrope, fn_combatLanding, and fn_createVehicleQRFBehaviour that I think it would probably be best to test and implement them separately.

jwoodruff40 committed:
Update fn_combatLanding.sqf
